File: app/src/main/java/com/nollpointer/dates/Date.java
package com.nollpointer.dates;

import android.os.Parcel;
import android.os.Parcelable;

public class Date implements Parcelable{
    private String date;
    private String event;
    private String request;
    private int type;

    private String month;

    public Date(String date, String event, String request, int type) {
        this.date = date;
        this.event = event;
        this.request = request;
        this.type = type;
    }

    public String getDate() {
        return date;
    }

    public void setDate(String date) {
        this.date = date;
    }

    public String getEvent() {
        return event;
    }

    public void setEvent(String event) {
        this.event = event;
    }

    public String getRequest() {
        return request;
    }

    public void setRequest(String request) {
        this.request = request;
    }

    public int getType() {
        return type;
    }

    public void setType(int type) {
        this.type = type;
    }

    public String getMonth() {
        return month;
    }

    public void setMonth(String month) {
        this.month = month;
    }

    public boolean containsMonth(){
        return month != null;
    }

    protected Date(Parcel in) {
        date = in.readString();
        event = in.readString();
        request = in.readString();
        type = in.readInt();
        month = in.readString();
    }

    public boolean isSameDate(Date date){
        return this.getDate().equals(date.getDate());
    }

    public boolean isContinuous (){
        return date.contains("-") || date.contains("–");
    }

    public boolean contains(String query){
        return date.toLowerCase().contains(query.toLowerCase()) || event.toLowerCase().contains(query.toLowerCase());
    }


    //Parcelable implementation

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int flags) {
        parcel.writeString(date);
        parcel.writeString(event);
        parcel.writeString(request);
        parcel.writeInt(type);
        parcel.writeString(month);
    }

    public static final Parcelable.Creator<Date> CREATOR = new Parcelable.Creator<Date>() {
        @Override
        public Date createFromParcel(Parcel in) {
            return new Date(in);
        }

        @Override
        public Date[] newArray(int size) {
            return new Date[size];
        }
    };

}

File: app/src/main/java/com/nollpointer/dates/model/Date.kt
package com.nollpointer.dates.model

import android.os.Parcelable
import kotlinx.parcelize.Parcelize
import java.util.*

/**
 * Модель Даты
 *
 * @author Onanov Aleksey (@onanov)
 */
@Parcelize
data class Date(
        var date: String = "",
        var event: String = "",
        var request: String = "",
        var type: Int = 1,
        var month: String = "#",
) : Parcelable, Comparable<Date> {

    override fun compareTo(other: Date): Int {
        val currentDate = date.toInt()
        val compareDate = other.date.toInt()
        return currentDate.compareTo(compareDate)
    }

    val dateFull: String
        get() = if (containsMonth) "$date, $month" else date

    val isContinuous: Boolean
        get() = date.contains("-") || date.contains("–") || date.contains("–") || date.contains(",") || date.contains("в")

    val containsMonth: Boolean
        get() = month != "#"

    operator fun contains(q: String): Boolean {
        val query = q.toLowerCase(Locale.ROOT)
        return date.toLowerCase(Locale.ROOT).contains(query) || event.toLowerCase(Locale.ROOT).contains(query) || containsMonth && month.toLowerCase(Locale.ROOT).contains(query)
    }

}